### ISSUE TITLE
Does confluencebuilder support Jinja 2? Getting error

### ISSUE THREAD

**Issue by binarylandscapes** [sphinx-contrib/confluencebuilder#440]
Running Sphinx v3.4.3
making output directory... done
running in headless mode, starting Xvfb
Xvfb is running on display :1
checking for /mnt/c/gitrepos/skd/mission-systems/communications-segment/network/sku_design/source/common_docs/references/refs.bib in bibtex cache... not found
parsing bibtex file /mnt/c/gitrepos/skd/mission-systems/communications-segment/network/sku_design/source/common_docs/references/refs.bib... parsed 2 entries
building [mo]: all of 0 po files
building [confluence]: all source files
updating environment: [new config] 24 added, 0 changed, 0 removed
reading sources... [100%] index
<unknown>:233: WARNING: TODO entry found: TODO: Need to find and integrate the SCADE Model Requirement Traceability Matrix
<unknown>:249: WARNING: TODO entry found: TODO: Need to find and integrate the SCADE Equipment Functional Allocation List
<unknown>:265: WARNING: TODO entry found: TODO: Need to find and integrate the SCADE Model Quality Checklist
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
(drawio) 'configs/drawings/arch_phys.drawio' -> 'drawio/800f0f4571dcd3d1bb449420f2f183ab02aec628/arch_phys.jpg'
(drawio) 'configs/drawings/arch_logical_air.drawio' -> 'drawio/b27b3597bfc07f83c87fcb2397884e0532e25e32/arch_logical_air.jpg'
(drawio) 'configs/drawings/arch_logical_ground.drawio' -> 'drawio/eb66a1f94ed0d5027c9a9ef78caf9b61dfbfed60/arch_logical_ground.jpg'
(drawio) 'configs/drawings/arch_dataflow.drawio' -> 'drawio/b64242a5de853c3f2e6f178bc72cede0ee28f259/arch_dataflow.jpg'

**Extension error:
Handler <function setup.<locals>.assetsDocTreeResolvedHook at 0x7fbf9d958af0> for event 'doctree-resolved' threw an exception (exception: 'PosixPath' object has no attribute 'startswith')
make: *** [Makefile:87: confluence] Error 2**

**Comment by binarylandscapes** [comment]
I run the following extensions:

extensions = [
    'sphinx.ext.autodoc',
    'sphinx.ext.intersphinx',
    'sphinx.ext.autosectionlabel',
    'sphinx.ext.todo',
    'sphinx.ext.coverage',
    'sphinx.ext.imgmath',
    'sphinx.ext.ifconfig',
    'sphinx.ext.viewcode',
    'sphinxcontrib.jinja',
    'sphinxcontrib.seqdiag',
    'sphinxcontrib.nwdiag',
    'sphinxcontrib.rackdiag',
    'sphinxcontrib.packetdiag',
    'sphinxcontrib.blockdiag',
    'sphinxcontrib.actdiag',
    'sphinx_git',
    'sphinxcontrib.plantuml',
    'sphinxcontrib.jupyter',
    'sphinxcontrib.ansibleautodoc',
    'sphinxcontrib.confluencebuilder',
    'sphinx.ext.extlinks',
    'sphinx.ext.githubpages',
    'sphinxcontrib.drawio',
    'sphinxdrawio.drawio_html',
    'sphinxcontrib.bibtex',
]

an I have a lot of Jinja templating fed by yaml as well

**Comment by kmitro** [comment]
I've got the same error with trying to implement support of publishing draw.io documents.

Extension error (sphinxcontrib.confluencebuilder):
Handler <function setup.<locals>.assetsDocTreeResolvedHook at 0x7f0fd5edf378> for event 'doctree-resolved' threw an exception (exception: 'PosixPath' object has no attribute 'startswith')


It looks like some kind of incompatibility for sphinxcontrib.drawio and sphinxcontrib.confluencebuilder.

Here is my list:

extensions = [
    'sphinx.ext.todo', 'recommonmark', 'sphinxcontrib.mermaid','sphinxcontrib.confluencebuilder', 'sphinxcontrib.drawio'
]

**Comment by jdknight** [comment]
Finally able to look at this now, two points I'd like to share from some recent investigation.

First impressions for this issue was assumed to be focused on Jinja templating. At this time, there is no stock support to process any custom templating with this extension. The builders provided in the `confluencebuilder` extension do not provide any internal templating support, let alone attempt to provide support for theme factories/templates/etc. Typically the ideal of theme management is managed on Confluence spaces (since Confluence document publishing does not dictate themes on an instance). However, it would be nice to provide somewhat of a crude support to process some sort of theming capability. Crudely injection this from a test environment does yield some template processing support:

```
from sphinx.theming import HTMLThemeFactory

def rstjinja(app, docname, source):
    if app.builder.format == 'confluence_storage':
        theme_factory = HTMLThemeFactory(app)
        app.builder.theme = theme_factory.create('alabaster')
        app.builder.theme_options = {}
        app.builder.create_template_bridge()
        app.builder.templates.init(app.builder, app.builder.theme)

        src = source[0]
        rendered = app.builder.templates.render_string(
            src, app.config.html_context
        )
        source[0] = rendered

def setup(app):
    app.connect("source-read", rstjinja)
```

But there is a lot to consider here. While we could add some basic template support, there can be complications having users inject raw storage format templates into documents (although that may be fine -- if a user wants to make custom storage format themes, why not). Something like this can be consider an enhancement and investigated.

---

Second thing, I overlooked what appears to be the root issue spawned in these comments. Indeed, there appears to be an issue with the integration of this extension and the `sphinxcontrib.drawio` extension. It appears the `uri` attributes for generates nodes are not string types by Python path types -- while may not be wrong, is it something that was not really considered by the `confluencebuilder` extension. Modifications can be made to better handle these scenarios. This is a bug and it should be fixed.

**Comment by jdknight** [comment]
For the errors reports in this issue, the merge of #473 should resolve issues when publishing with images generated by the `sphinxcontrib.drawio` extension. This change is available on the main development branch and will be made available next stable release (v1.6).

**Comment by binarylandscapes** [comment]
Thanks for looking into this, yes I by no means expected tons of work for templating but my root issue was with draw.io. I have found there are some limitations with jinja and the output but in my use case jinja is used for text content not page templating in that since, so that should be handled in Sphinx prior to the produced rst output that I assume is what is feed into confluencebuilder. but it wouldnt build at all with drawio as part of document. So I will retest when 1.6 is released and provide feedback here

**Comment by jdknight** [comment]
An update.

## sphinxcontrib.drawio

Support for this extension is possible. Testing against the latest version of the drawio extension, a workaround is needed for the current v1.6 release of the Confluence builder extension. Including the following will ensure the DrawIONode is ignored:

```
confluence_adv_ignore_nodes = [
    'DrawIONode',
]
```

I'm not sure off hand why the node is still provided to the translator after the drawio extension prepares an image reference. But ignoring the node type should still generate images for a documentation set. Adding this workaround will not be needed for a v1.7 release (current development branch) as unknown nodes will only generate a warning now (instead of an error).

Note that the prospect v1.7 release of the Confluence builder extension does provide improved SVG support. If using drawio's to render SVGs and sizing options result in attachment errors, this should be resolved in a future release.

## sphinxcontrib.jinja

Performing some tests today does show that it is possible to use this extension with the Confluence builder extension. Although when testing, there seemed to be issues with a couple of issues with the jinja extension (jinja template files would not always load right, filters did not appear to be loaded into the Jinja environment, etc.). That being said, when able to create a documentation set using the Jinja extension that worked with an `html` builder, a `confluence` generated page appeared to work as well.

Attaching the example documentation:

[conf.py](https://github.com/sphinx-contrib/confluencebuilder/files/7536170/conf.py.txt), [index.rst](https://github.com/sphinx-contrib/confluencebuilder/files/7536175/index.rst.txt), [test.json](https://github.com/sphinx-contrib/confluencebuilder/files/7536176/test.json.txt)

---


Since the primary issue outlined in this thread was resolved with the changes introduced in #473, and this fix is available in the existing v1.6 release, marking this issue as closed. If anyone who does use the Jinja extension with Confluence builder and observes and issue with the content generated by this extension, please feel free to create a new issue with the specifics.
